mpqp/execution/providers/google.py

```diff
@@ -18,6 +20,96 @@
+        noises: A list of noise models to apply to the circuit.
+        nb_qubits: The number of qubits in the circuit.
+
+    Returns:
+        A new circuit with the noise operations applied.
+    """
+    from cirq.ops.identity import IdentityGate
+    from cirq.ops.measurement_gate import MeasurementGate
+    from cirq.ops.raw_types import Gate, Operation
+
+    qubit_list = sorted(cirq_circuit.all_qubits())
+    noisy_moments = []
+
+    for moment in cirq_circuit:
+        moment_ops = list(moment.operations)
+        noisy_moments.append(Moment(moment_ops))
+
+        qubit_noise_op: list[list[Operation]] = [[] for _ in range(nb_qubits)]
+
+        for op in moment_ops:
+            if isinstance(op.gate, (MeasurementGate, IdentityGate)):
+                continue
+
+            for noise in reversed(noises):
+                cirq_noise = noise.to_other_language(Language.CIRQ)
+                if TYPE_CHECKING:
+                    assert isinstance(cirq_noise, Gate)
+
+                noise_dimension = getattr(noise, "dimension", 1)
+
+                if not noise.targets or len(noise.targets) == nb_qubits:
+                    target_qubits = qubit_list
+                else:
+                    target_qubits = [qubit_list[i] for i in noise.targets]
+
+                allowed_gates = set()
+                if noise.gates:
+                    for gate in noise.gates:
+                        gate_obj = (
+                            gate.cirq_gate if hasattr(gate, "cirq_gate") else gate
+                        )
+                        gate_cls = (
+                            type(gate_obj)
+                            if not isinstance(gate_obj, type)
+                            else gate_obj
+                        )
+                        allowed_gates.add(gate_cls)
+
+                if noise_dimension == 2 and len(op.qubits) == 2:
+                    if all(q in target_qubits for q in op.qubits):
+                        if not allowed_gates or isinstance(
+                            op.gate, tuple(allowed_gates)
+                        ):
+                            noisy_gate = cirq_noise.on(*op.qubits)
+                            noisy_moments.append(Moment([noisy_gate]))
+
+                elif noise_dimension == 1:
+                    for q in op.qubits:
+                        if q in target_qubits:
+                            if not allowed_gates or isinstance(
+                                op.gate, tuple(allowed_gates)
+                            ):
+                                qubit_index = qubit_list.index(q)
+                                qubit_noise_op[qubit_index].append(cirq_noise.on(q))
+
+        while any(qubit_noise_op):
+            current_moment_ops = []
+            for qubit_ops in qubit_noise_op:
+                if qubit_ops:
+                    current_moment_ops.append(qubit_ops.pop(0))
+            noisy_moments.append(Moment(current_moment_ops))
```

Comment:
ok, I understood the logic now, here is a proposition of make it easier (to me) to understand at a glance:
```suggestion
        moments = []
        for moment_index in max(len(ops) for ops in qubit_noise_op):
            moments.append([ops[moment_index] for ops in qubit_noise_op if len(ops) > moment_index])
```